CRM-21244 Enhancements to "FROM email addresses"

mattwire opened the pull request:
Overview
----------------------------------------
This was developed from a client requirement to only allow sending email from the organisation from addresses, not from individual logged in account email addresses.  Initially I used the extension https://github.com/CiviCooP/org.civicoop.protectfromaddress to do this.  However, it became clear that this only works on certain forms.  Extending the list of forms (to around 20) helps in most cases but not all.  I then investigated the code in core and discovered a number of places where the same code was basically duplicated instead of being shared.  So this patch does the following:
1) Clean up all duplicated "FROM email address" code into a shared function in CRM_Core_BAO_Email.
2) Add a setting in that shared function that decides whether to load logged in contact email addresses or not.  (Currently setting is located in "Outbound Email", but don't really mind where it goes if there's a better place).
3) Remove "FROM email address" setting from organizational contact settings as it just gets/sets the first one listed in "FROM email addresses" settings page and it's always confused me as to what the difference was (there is no difference).
4) Add "FROM email address" select to the PDF letter task (for consistency with all other tasks, previously this one was hardcoded to take the first address in the generated array of FROM addresses).

---

 * [CRM-21244: Enhancements to "FROM email addresses"](https://issues.civicrm.org/jira/browse/CRM-21244)

seamuslee001 commented:
jenkins re test this please

seamuslee001 commented:
@mattwire I like the direction this is heading in, The test failures look related tho

mattwire commented:
Jenkins test this please

seamuslee001 commented:
jenkins retest this please

seamuslee001 commented:
Jenkins re test this please

seamuslee001 commented:
Initial error was `Fatal error:  Allowed memory size of 2147483648 bytes exhausted (tried to allocate 523800 bytes) in /home/jenkins/buildkit/build/core-11047-8p09g/sites/all/modules/civicrm/CRM/Core/BAO/Domain.php on line 173`

seamuslee001 commented on CRM/Core/BAO/Domain.php:
```diff
@@ -169,9 +169,13 @@ public static function multipleDomains() {
-  public static function getNameAndEmail($skipFatal = FALSE) {
-    $fromEmailAddress = CRM_Core_OptionGroup::values('from_email_address', NULL, NULL, NULL, ' AND is_default = 1');
+  public static function getNameAndEmail($skipFatal = FALSE, $returnString = FALSE) {
+    $fromEmailAddress = CRM_Core_BAO_Domain::getNameAndEmail(TRUE, TRUE);
```
this is creating a nice infinite loop

mattwire commented:
jenkins retest this please

mattwire commented:
@eileenmcnaughton Hey, any idea why the AuthorizeIPN tests are failing?  I don't see how the body of an email could be affected by my changes. @seamuslee001 said you might know :-)

JoeMurray commented:
I don't mind seeing code cleanup to standardize similar code into one place. However, I think I'd prefer to see the setting and restriction to using emails from same domain as site (if that is what is contemplated) remain in an extension. We want to reduce the size and complexity of core by breaking functionality in core out into new extensions, not fold good and useful extensions into core.

Also, is the From email address mentioned above not used in some parts of CiviEvent etc on info pages?

eileenmcnaughton commented:
Hmm - this does seem to reduce complexity rather than increase it. It does add a setting - which is obviously an orange flag for 'could this be an extension'.  In this case the setting only affects the contents of the options for one select box & doesn't affect post-process & it feels very clear & logical so I am not opposed to it, although  I suspect the clean-up would make it fairly easy to do as a hook instead.

 I would note that poorly thought out hooks are at least as bad for code maintainability as additional settings.

If we do put this setting in core I think it would make more sense as a permission. 

Note that the proposed setting / permission is whether the user's email will be populated as an option into the from field for emails (if not then just the domain email will remain)

mattwire commented:
@eileenmcnaughton @JoeMurray 

> Note that the proposed setting / permission is whether the user's email will be populated as an option into the from field for emails (if not then just the domain email will remain).

Correct. The setting doesn't affect the domain emails.  I don't mind if it's a setting or a permission - not sure that a hook is sensible here as I don't think it's useful for anything else that can't already be configured via the UI but I could be persuaded if there was a realistic use-case.

This part of the code was a complete mess, luckily well covered by unit tests so this patch goes a long way to improving maintainability and reducing complexity in the future.  It also "removes" a setting from the domain settings page which actually duplicates what's already on the from addresses page.

eileenmcnaughton commented:
@seamuslee001 are you able to look through these code changes?

seamuslee001 commented:
@eileenmcnaughton i don't know enough about the way the Authorized.net tests are meant to be working, so I can take a look at the rest but don't know how to fix the test failures

JKingsnorth commented:
Made some progress on this, will provide a writeup soon.

JKingsnorth commented:
Ready to take a trip down the rabbit hole? OK, here's why the test is failing @mattwire ...

It's due to the refactoring of CRM_Contact_BAO_Contact_Location::getEmailDetails.

The only times the second two parameters (isPrimary, locationTypeID) are used is in:
- \CRM_Contribute_BAO_ContributionPage::recurringNotify
- \CRM_Contribute_BAO_ContributionPage::sendMail

In these cases the function can actually return **multiple** email addresses. The newly refactored method has 'limit 1' on the API call.

Additionally, in the original function isPrimary = FALSE doesn't mean "don't return the primary result" - it means "don't limit the search to just the primary result".

Quick fix: Get the getEmailDetails function to return like it used to (with the improvements from refactoring). Here's a patch that will achieve this: https://github.com/civicrm/civicrm-core/pull/11164/commits/631768ecffb1d21867baf428f58d63df4a42675b (https://github.com/civicrm/civicrm-core/pull/11164)

Follow-up:
- Add a test to cover getEmailDetails and the types of return we are expecting
- Figure out why 'recurringNotify' is looking for non-primary email addresses??
- sendMail is looking for the 'business' email for a contact, probably fair enough?

JKingsnorth commented:
p.s. +144 −271 < Nice

monishdeb commented on CRM/Admin/Form/Setting/Smtp.php:
```diff
@@ -79,6 +80,14 @@ public function postProcess() {
+    else {
+      Civi::settings()->set('allow_mail_from_logged_in_contact', FALSE);
+    }
+    unset($formValues['allow_mail_from_logged_in_contact']);
```
This can be further optimised into 
```patch
diff --git a/CRM/Admin/Form/Setting/Smtp.php b/CRM/Admin/Form/Setting/Smtp.php
index e3219e8..d05ff48 100644
--- a/CRM/Admin/Form/Setting/Smtp.php
+++ b/CRM/Admin/Form/Setting/Smtp.php
@@ -80,12 +80,7 @@ class CRM_Admin_Form_Setting_Smtp extends CRM_Admin_Form_Setting {
 
     $formValues = $this->controller->exportValues($this->_name);
 
-    if (!empty($formValues['allow_mail_from_logged_in_contact'])) {
-      Civi::settings()->set('allow_mail_from_logged_in_contact', TRUE);
-    }
-    else {
-      Civi::settings()->set('allow_mail_from_logged_in_contact', FALSE);
-    }
+    Civi::settings()->set('allow_mail_from_logged_in_contact', (!empty($formValues['allow_mail_from_logged_in_contact'])));
```

monishdeb commented on CRM/Admin/Form/Setting/Smtp.php:
```diff
@@ -238,6 +247,8 @@ public function setDefaultValues() {
       $this->_defaults = array();
 
+      $this->_defaults['allow_mail_from_logged_in_contact'] = Civi::settings()->get('allow_mail_from_logged_in_contact');
+
```
Put this assignment below just before return because at L254 it overrides the default values and the value for this new setting is no longer set as default.

monishdeb commented on CRM/Core/BAO/Email.php:
```diff
@@ -276,29 +276,41 @@ public static function holdEmail(&$email) {
+   */
+  public static function domainEmails() {
+    $domainEmails = array();
+    $domainFrom = CRM_Core_OptionGroup::values('from_email_address');
```
Need array casting as 
```
$domainFrom = (array) CRM_Core_OptionGroup::values('from_email_address');
``` to handle situation if it returns NULL;

monishdeb commented:
I have tested the patch and it works for me. Here's what I have noticed:
1. The new setting ```allow_mail_from_logged_in_contact``` is available in 'Outbound Mail Setting' page:
![screen shot 2017-11-02 at 3 27 15 am](https://user-images.githubusercontent.com/3735621/32300044-d24be8aa-bf7d-11e7-8d87-5a39e5f31222.png)
It looks a bit odd though to show checkbox in label setion and the field is in middle not on left.
2. CRM_Contact_BAO_Contact_Location::getEmails(...) is modified to use Email.Get API rather then SQL queries and is responding well.
3. Two fields removed from ```Organization Address and Contact Info``` form, which are 'From Name' and 'From Email address'. Which makes sense as we are fetching From email address either of logged in contact OR set option ```from_email_address ``` values. And CRM_Core_BAO_Email::getFromEmail() decide that on basis of new setting value. No unintended issue found. 
4. Necessary refactoring changes made in ```CRM_Contact_Form_Task_EmailCommon::preProcessFromAddress($form)``` and tested it's effect on different tasks (Print Invoice, Send Email etc.). No unintended regression found.
5. Link used in description or help text, in order to add/edit From email address changed to use 'from_email_address' option value page.
6. Additional changes to provide helpicon to backoffice form

Overall it's working fine. However, as per code pov I have made few comments. Also can you squash the commits into 1 via ```git rebase -i``` ?

mattwire commented:
@monishdeb Thankyou for the review.  I think I've addressed all your comments.
The settings now looks like this (I've added a crmRegion as the sparkpost email extension currently overrides the whole page, with the crmRegion it can just replace the mailer configuration):
![localhost_8001_civicrm_admin_setting_smtp_reset 1](https://user-images.githubusercontent.com/2052161/32466158-e1815558-c33d-11e7-9816-f9abc8b77d18.png)

eileenmcnaughton commented:
While it's not a blocker to this, I do rather aspire to moving settings templates to being more like the Xero version where they just render the assigned fields based on the metadata.

https://github.com/eileenmcnaughton/nz.co.fuzion.civixero/blob/master/templates/CRM/Civixero/Form/XeroSettings.tpl

mattwire commented:
@monishdeb Could you confirm if you are happy that I've addressed all your review comments?

mattwire commented:
@eileenmcnaughton @totten Either of you got time to review and/or merge this?

mlutfy commented:
Does this affect "Send to a friend"? (I'm OK with considering that a different issue, just putting it out there) :)

mattwire commented:
@mlutfy Where is the "Send to a friend?" functionality?  I can check if this could have an impact.

mlutfy commented:
@mattwire On Contribution Pages settings, there is a tab "Tell a Friend", which can be enabled. However, I always considered this feature to be rather broken, since CiviCRM emails using the user's email, which is wrong (fails SPF/DKIM).

mattwire commented:
@mlutfy It will have no affect on "Tell a Friend" which always pulls the email address of the contact making the contribution via CRM_Core_DAO_Email (https://github.com/civicrm/civicrm-core/blob/master/CRM/Friend/Form.php#L194).  Something should certainly be done about that, as it will fail SPF but it's out of the scope of this PR.

yashodha commented:
![from_address](https://user-images.githubusercontent.com/3455173/33211080-de1f56d6-d142-11e7-80a2-006489ca18b1.png)
I applied the patch and it seems to be adding the "From Address" twice on Thank you task

mattwire commented:
@yashodha I've just pushed a fix that should fix your issue.  Please can you test again?

yashodha commented:
@mattwire looks good now and works fine ! Only thing I found was help message was missing on clicking the info 
![from_missing_help_text](https://user-images.githubusercontent.com/3455173/33252586-1b149c30-d365-11e7-9b03-c747b96123ac.png)

mattwire commented:
@yashodha Thankyou for testing.  The missing help should now be fixed.

yashodha commented:
@mattwire tested, works fine now on all screens

eileenmcnaughton commented:
A lot of people have touched this so it's hard to know if it's ready to merge or not - @yashodha were you going to merge it?

yashodha commented:
@mattwire on further testing, I found we are not sending the from address to the post if we change the email address 
Can you apply this patch which should make it work
```php
diff --git a/CRM/Contribute/Form/Task/PDFLetterCommon.php b/CRM/Contribute/Form/Task/PDFLetterCommon.php
index 44aaf47..faa041e 100644
--- a/CRM/Contribute/Form/Task/PDFLetterCommon.php
+++ b/CRM/Contribute/Form/Task/PDFLetterCommon.php
@@ -23,6 +23,7 @@ class CRM_Contribute_Form_Task_PDFLetterCommon extends CRM_Contact_Form_Task_PDF
       $returnProperties['email'] = $returnProperties['on_hold'] = $returnProperties['is_deceased'] = $returnProperties['do_not_email'] = 1;
       $emailParams = array(
         'subject' => $formValues['subject'],
+        'from' => $formValues['from_email_address'],
       );

```

yashodha commented:
@eileenmcnaughton submitted more feedback :)

eileenmcnaughton commented:
thanks @yashodha - when you are happy with this can you please merge it?

yashodha commented:
@eileenmcnaughton sure thing :)

yashodha commented:
@mattwire another thing that I encountered another bug. Can't print/merge document with as from address has been added in CRM/Contact/Form/Task/PDFLetterCommon.php https://github.com/civicrm/civicrm-core/pull/11047/files#diff-838ac329f5c830ce93bc0d91580ae1a7R108
This will happen for any of the event, contact tasks for _Print/Merge Document_
We need a check or something like

```
diff --git a/CRM/Contact/Form/Task/PDFLetterCommon.php b/CRM/Contact/Form/Task/PDFLetterCommon.php
index 4492bce..2cf8728 100644
--- a/CRM/Contact/Form/Task/PDFLetterCommon.php
+++ b/CRM/Contact/Form/Task/PDFLetterCommon.php
@@ -101,8 +101,9 @@ class CRM_Contact_Form_Task_PDFLetterCommon {
       FALSE
     );
 
-    $form->add('select', 'from_email_address', ts('From Email Address'), $form->_fromEmails, TRUE);
-
+    if (get_class($form) == 'CRM_Contribute_Form_Task_PDFLetter') {
+      $form->add('select', 'from_email_address', ts('From Email Address'), $form->_fromEmails, TRUE);
+    }
     $form->add('static', 'pdf_format_header', NULL, ts('Page Format: %1', array(1 => '<span class="pdf-format-header-label"></span>')));
     $form->addSelect('format_id', array(
       'label' => ts('Select Format'),


```

mattwire commented:
@yashodha Thankyou.  I've just pushed a fix that should address both the bugs you found. For the CRM_Contribute_Form_Task_PDFLetter form I've added a new buildForm method as that is the only PDF form that needs the from_email_address.

eileenmcnaughton commented:
test failure relates

yashodha commented:
@mattwire Can you please address this https://test.civicrm.org/job/CiviCRM-Core-PR/18269/testReport/junit/(root)/CRM_Contribute_Form_Task_PDFLetterCommonTest/testPostProcessGroupByContact/

We should be using 
```php
diff --git a/CRM/Contribute/Form/Task/PDFLetterCommon.php b/CRM/Contribute/Form/Task/PDFLetterCommon.php
index 44aaf47..faa041e 100644
--- a/CRM/Contribute/Form/Task/PDFLetterCommon.php
+++ b/CRM/Contribute/Form/Task/PDFLetterCommon.php
@@ -23,6 +23,7 @@ class CRM_Contribute_Form_Task_PDFLetterCommon extends CRM_Contact_Form_Task_PDF
       $returnProperties['email'] = $returnProperties['on_hold'] = $returnProperties['is_deceased'] = $returnProperties['do_not_email'] = 1;
       $emailParams = array(
         'subject' => $formValues['subject'],
+        'from' => CRM_Utils_Array::value('from_email_address', $formValues),
       );
```

mattwire commented:
@yashodha Thanks, I've updated that line and it's passing now.

yashodha commented:
@mattwire The field is not showing now
![from_addr](https://user-images.githubusercontent.com/3455173/33611954-1e54e692-d9f6-11e7-8d4f-006f86f3ac0a.png)

mattwire commented:
@yashodha I think I've finally got it.  The fix I put in before was ignored as we weren't calling CRM_Contribute_Task_PDFLetterCommon but calling CRM_Contact_Task_PDFLetterCommon directly.

yashodha commented:
@mattwire works fine. Merging this now.

eileenmcnaughton commented:
Yay, well done @yashodha @mattwire

Stoob commented:
Hey @eileenmcnaughton @seamuslee001 and @JoeMurray I think Matt did some nice cleanup here, but can you please explain what happened on the changing of default handling of 'from' email on a 1:1 or 'small batch' context?   As you know this is a common use case and most importantly it's how it's documented to work.  Respecting the needs of others (and their clients), please know this change has recently caused confusion and frustration others - who want it changed back (CRM-21244). 

![emeail-select](https://user-images.githubusercontent.com/1683473/38156839-1ed94546-3436-11e8-9627-d1033f6537f3.png)

seamuslee001 commented:
@Stoob as AFAIK the main change was that it may have excluded logged in users's email address by default. This can be turned back on in the SMTP settings by setting allow_mail_from_logged_in_contact setting. IMO i think its better to update the documentation here about the setting. I'm not sure why we should be allowing emails getting sent from potentially random domains. 

One potential solution here maybe that the setting is automatically turned on and users have to manually turn it off for current installs

Stoob commented:
@seamuslee001  thanks but that setting is already on, and the logged-in-user's email address is available from the selection, BUT logged-in-user's email it isn't the default.  Logged-in-user's email address _was_ the default, and the documentation says it is the default, but it's not the default anymore.  Instead, the default global 'from' address is the default.  It could be a matter of simply switching the order in which the emails are displayed/selected, with the logged-in users as the top/default.

Do I need to go into the use case here, or share some examples on how this has caused havok/confusion for people were unaccustomed to having to verify from the 'from' email address was theirs every time they send an email?

seamuslee001 commented:
@Stoob check https://github.com/civicrm/civicrm-core/pull/11905 to see if that works for you

JoeMurray commented:
Good point @stoob. I see my one comment favoured moving any functional change into an extension rather than having it change core.

Stoob commented:
Thank you for your quick work Seamus!  Here's another troublesome bit, which occurred just Wed after the upgrade from 4.7.26 to 4.7.31.  Please note: no server configurations have changed during this time.

When I sent an email in the following way contact to contact, 1to1:
![single-domain-non-website](https://user-images.githubusercontent.com/1683473/38159888-6f1a39ee-3467-11e8-8c60-454af62f1822.png)

This is the result:
![arrival](https://user-images.githubusercontent.com/1683473/38159897-a785a930-3467-11e8-9356-20610aeece22.png)

Yet when I sent a test via civicrm/admin/setting/smtp or CiviMailBulk the email arrives with a normal 'from' line.
![arrival3](https://user-images.githubusercontent.com/1683473/38159913-3ebaeab8-3468-11e8-9782-ddf7eb49c5e0.png)
![arrival2](https://user-images.githubusercontent.com/1683473/38159914-3ee95f6a-3468-11e8-8340-9dcaf2c599a2.png)

Thoughts why?

Stoob commented:
@seamuslee001 you are correct.  When I choose a 'from' email that is not attached to the currently logged-in contact the FROM line works perfectly.

seamuslee001 commented:
@Stoob i have just pushed a 2nd fix up to my PR to try and fix that issue as well.

Stoob commented:
Hurray!   Your 2nd patch works, from link from a contact-specific email works.  You are awesome!

eileenmcnaughton commented:
Thanks @stoob for giving useful explanations of the issue & prompt testing of fixes & @seamuslee001 for fixing it. Great work!
